### PR TITLE
Support for channel based properties

### LINKED ISSUE #185: Channels Context Manager for properties and functions (“with session.channels(‘0’) as c:”)
On multi-channel sessions (think NI-DCPower), often times customer wants to call functions or set/get attributes on specific channel(s).

We want to create a context manager that allows doing exactly this without having to pass the same channel parameter in multiple calls. In LabVIEW, you get equivalent functionality for properties by using the Active Channel input in the property node.

    my_session.do_foo("0,1", 5)
    b = my_session.get_foo("0,1")
    # One way, proposed in https://github.com/ni/nimi-python/issues/187
    my_session.a_property["0,1"] = 5
    # Another potential way
    my_session.set("a_property", "0,1", 5)
    # getting is left as an exercise to the reader.

Could instead be achieved as:

    with my_session.channel("0,1") as channel_context:
        channel_context.a_property = 5
        value = channel_context.another_property

### LINKED ISSUE #187: Add channels for properties on session object
Example:

`session.voltage_level[‘0,1’] = 5 `

Would set voltage_level property to 5 on channels 0, 1.

### PR DESCRIPTION
**Pull request by marcoskirsch**
[X] This contribution adheres to [CONTRIBUTING.md]

### What does this Pull Request accomplish?

* Adds API support for setting and getting channel-based properties.
* Paves the way for how we want to handle the channel parameter from out C API.
* Fixes #185 
* Fixes #187 

Channel based properties can be used in two ways:

```
# sets voltage level in channel 2 to 5 volts.
session['2'].voltage_level = 5
session['2']. output_enabled = 5
```

```
# using a context manager.
with session['2'] as c2:
    c2.voltage_level = 5
    c2.output_enabled = True
```

The same syntax also applies to all repeated capabilities, not only channels:

```
session['ScriptTrigger0'].source_terminal = "PXI_Trig5"
```

### Why should this Pull Request be merged?

This is new functionality for the Python APIs. Required by most of the Modular Instruments drivers.

### What testing has been done?

* Unit tests were added.
* Some local testing with NI-SWITCH was performed.

### CODE REVIEW

File: build/templates/session.py.mako [hunk 1]
```diff
@@ -158,3 +138,52 @@ init_call_params = helper.get_params_snippet(init_function, helper.ParamListType
 % endif
 % endfor
 
+
+class _RepeatedCapability(_SessionBase):
+    '''Allows for setting/getting property values for specific channels in your session.'''
```
Comment (marcoskirsch) [inline build/templates/session.py.mako]: improve wording.

File: build/templates/session.py.mako [hunk 2]
```diff
@@ -158,3 +138,52 @@ init_call_params = helper.get_params_snippet(init_function, helper.ParamListType
...
+
+    def __init__(self, vi, channel):
+        super(_RepeatedCapability, self).__init__(repeated_capability=channel)
```
Comment (marcoskirsch) [inline build/templates/session.py.mako]: don't use 'channel'

File: build/templates/session.py.mako [hunk 3]
```diff
@@ -158,3 +138,52 @@ init_call_params = helper.get_params_snippet(init_function, helper.ParamListType
...
+        self.vi = vi
```
Comment (marcoskirsch) [inline build/templates/session.py.mako]: use session_handle_parameter_name

File: build/templates/session.py.mako [hunk 4]
```diff
@@ -158,3 +138,52 @@ init_call_params = helper.get_params_snippet(init_function, helper.ParamListType
...
+
+    def __getitem__(self, repeated_capability):
+        '''Set/get properties or call methods with a repeated capabiilty (i.e. channels)'''
+        return _RepeatedCapability(self.vi, repeated_capability)
+
+    def initiate(self):
+        return ${session_context_manager}(self)
+
+    def close(self):
+        # TODO(marcoskirsch): Should we raise an exception on double close? Look at what File does.
```
Comment (marcoskirsch) [inline build/templates/session.py.mako]: remove, we want to log a warning

File: build/templates/session.py.mako [hunk 5]
```diff
@@ -158,3 +138,52 @@ init_call_params = helper.get_params_snippet(init_function, helper.ParamListType
...
+        try:
+            self._close()
+        except errors.Error:
+            # TODO(marcoskirsch): This will occur when session is "stolen". Change to log instead
+            print("Failed to close session.")
+        self.vi = 0
```
Comment (marcoskirsch) [inline build/templates/session.py.mako]: use self.${config['session_handle_parameter_name']}

File: generated/nifake/tests/test_session.py [hunk 6]
```diff
@@ -49,7 +49,7 @@ def test_close(self):
 
     def test_session_context_manager(self):
         with nifake.Session('dev1') as session:
-            assert(session.vi == SESSION_NUM_FOR_TEST)
+            assert session.vi == SESSION_NUM_FOR_TEST
```
Comment (marcoskirsch) [inline generated/nifake/tests/test_session.py]: move all this cleanup to own PR

File: generated/nimodinst/attributes.py [hunk 7]
```diff
@@ -5,15 +5,14 @@
...
 
-    def __init__(self, attribute_id, default_channel=''):
+    def __init__(self, attribute_id):
         self._attribute_id = attribute_id
-        self._default_channel = default_channel
 
-    def __get__(self, obj, objtype):
-        return self.get(obj, self._default_channel)
+    def __get__(self, session, session_type):
+        return self.get(session, session._repeated_capability)
```
Comment (marcoskirsch) [inline generated/nimodinst/attributes.py]: This probably breaks niModInst.
We will know once the PR is posted and nimi-bot does its thing.

File: src/niswitch/examples/niswitch_channel_properties.py [hunk 8]
```diff
@@ -0,0 +1,17 @@
+#!/usr/bin/python
```
Comment (marcoskirsch) [inline src/niswitch/examples/niswitch_channel_properties.py]: This example probably should be removed.
We will add an example that shows this when we have something compelling to show (i.e. NI-DCPower).
@Fladolcetta what do you think? Ir is there something interesting to do with NI-SWITCH?

File: src/niswitch/examples/niswitch_channel_properties.py [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+import niswitch
+import sys
+
+parser = argparse.ArgumentParser(description='TODO: write something.', formatter_class=argparse.ArgumentDefaultsHelpFormatter)
+parser.add_argument('-n', '--name', default='PXI1Slot2', help='Resource name of a National Instruments Switch.')
+args = parser.parse_args()
+
+try:
+    with niswitch.Session(args.name) as session:
+        with session.channel("Marcos") as chan:
```
Comment (marcoskirsch) [inline src/niswitch/examples/niswitch_channel_properties.py]: example needs retrofitting anyway since this syntax changed.

File: build/templates/session.py.mako [hunk 10]
```diff
@@ -158,3 +138,52 @@ init_call_params = helper.get_params_snippet(init_function, helper.ParamListType
 % endif
 % endfor
 
+
+class _RepeatedCapability(_SessionBase):
+    '''Allows for setting/getting property values for specific channels in your session.'''
+
+    def __init__(self, vi, channel):
```
Comment (texasaggie97) [inline build/templates/session.py.mako]: Use repeated_capability instead of channel

File: build/templates/session.py.mako (hunk 3 above)
Comment (texasaggie97) [inline build/templates/session.py.mako]: Use ${config['session_handle_parameter_name']} instead of vi

File: build/templates/session.py.mako [hunk 11]
```diff
@@ -158,3 +138,52 @@ init_call_params = helper.get_params_snippet(init_function, helper.ParamListType
...
+        self._is_frozen = True
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.close()
+
+    def __getitem__(self, repeated_capability):
+        '''Set/get properties or call methods with a repeated capabiilty (i.e. channels)'''
```
Comment (texasaggie97) [inline build/templates/session.py.mako]: Typo

File: build/templates/session.py.mako (hunk 5 above)
Comment (texasaggie97) [inline build/templates/session.py.mako]: Use ${config['session_handle_parameter_name']} instead of vi

File: generated/nifake/tests/test_session.py [hunk 12]
```diff
@@ -351,6 +345,25 @@ def test_error_get_vi_real64_attribute(self):
                 assert e.code == test_error_code
                 assert e.description == test_error_desc
 
+    def test_get_channel_based_attribute(self):
```
Comment (texasaggie97) [inline generated/nifake/tests/test_session.py]: In this test you get, in the context manager test, you set.

I think you should test the other options as well: set without context manager and get with it.

File: generated/nifake/tests/test_session.py [hunk 13]
```diff
@@ -351,6 +345,25 @@ def test_error_get_vi_real64_attribute(self):
...
+        self.patched_library.niFake_GetAttributeViReal64.side_effect = self.side_effects_helper.niFake_GetAttributeViReal64
+        pi = 3.14159
+        self.side_effects_helper['GetAttributeViReal64']['attributeValue'] = pi
+        with nifake.Session('dev1') as session:
+            attr_double = session['0,1'].read_write_double
+            assert attr_double == pi
+            self.patched_library.niFake_GetAttributeViReal64.assert_called_once_with(SESSION_NUM_FOR_TEST, b'0,1', 1000001, ANY)
+
+    def test_channel_based_attribute_with_context_manager(self):
```
Comment (texasaggie97) [inline generated/nifake/tests/test_session.py]: Add 'set' to test name

File: src/niswitch/examples/niswitch_channel_properties.py [hunk 14]
```diff
@@ -0,0 +1,17 @@
...
+            print('channel {0} bandwidth is {1}'.format(chan, chan.bandwidth))
```
Comment (texasaggie97) [inline src/niswitch/examples/niswitch_channel_properties.py]: If you are going to use the repeated capability object in format, then you should add with \_\_format\_\_ or \_\_repr\_\_ or \_\_str\_\_ to the _RepeatedCapabilities class

File: src/niswitch/examples/niswitch_channel_properties.py (hunk 9 above)
Comment (texasaggie97) [inline src/niswitch/examples/niswitch_channel_properties.py]: Which switch has a "Marcos" channel?

File: build/templates/session.py.mako (hunk 11 above)
Comment (marcoskirsch) [reply build/templates/session.py.mako]: fixed

File: generated/nifake/tests/test_session.py (hunk 12 above)
Comment (marcoskirsch) [reply generated/nifake/tests/test_session.py]: I will expand on these tests

File: src/niswitch/examples/niswitch_channel_properties.py (hunk 9 above)
Comment (marcoskirsch) [reply src/niswitch/examples/niswitch_channel_properties.py]: This was written for local testing and I didn't mean to commit. I don't think the NI-SWITCH examples are that interesting for channel-based stuff so I'm leaning towards removing. We should write something good for NI-DCPower.
Thoughts?

File: src/niswitch/examples/niswitch_channel_properties.py (hunk 14 above)
Comment (marcoskirsch) [reply src/niswitch/examples/niswitch_channel_properties.py]: I had __repr__() for a bit, but there were some added trickiness so I removed and filed #328.
Need to delete this example.

File: generated/nifake/tests/test_session.py (hunk 6 above)
Comment (marcoskirsch) [reply generated/nifake/tests/test_session.py]: Done in #335

File: src/niswitch/examples/niswitch_channel_properties.py (hunk 9 above)
Comment (texasaggie97) [reply src/niswitch/examples/niswitch_channel_properties.py]: I think that is reasonable. So remove this example

File: src/niswitch/examples/niswitch_channel_properties.py (hunk 8 above)
Comment (marcoskirsch) [reply src/niswitch/examples/niswitch_channel_properties.py]: Removed

File: src/niswitch/examples/niswitch_channel_properties.py (hunk 9 above)
Comment (marcoskirsch) [reply src/niswitch/examples/niswitch_channel_properties.py]: I have removed this example and expanded src/niswitch/examples/niswitch_get_device_info.py

File: src/niswitch/examples/niswitch_get_device_info.py [hunk 15]
```diff
@@ -4,26 +4,44 @@
...
-        if args.information:
+
+        if args.device_info:
+            row_format = '{:<18}' * (2)
+            print(row_format.format('Device Name: ', session.io_resource_descriptor))
+            print(row_format.format('Device Model: ', session.instrument_model))
+            print(row_format.format('Driver Revision: ', session.specific_driver_revision))
+            print(row_format.format('Channel count: ', session.channel_count))
+            print(row_format.format('Relay count: ', session.number_of_relays))
+            print('Relay Info:')
```
Comment (Fladolcetta) [inline src/niswitch/examples/niswitch_get_device_info.py]: Dont print relay info as this is no longer printing relay info immediately afterwards. We can likely just delete this.

File: src/niswitch/examples/niswitch_get_device_info.py [hunk 16]
```diff
@@ -4,26 +4,44 @@
...
+                with session[channel_name] as channel:
+                    print(row_format.format(i, channel_name, str(channel.is_configuration_channel), str(channel.is_source_channel)))
+
+        if args.relay_info:
             row_format = '{:6}' + ' ' * 12 + '{:<15}{:<22}{:6}'
             print(row_format.format('Number', 'Name', 'Position', 'Count'))
             for i in range(1, session.number_of_relays + 1):
                 relay_name = session.get_relay_name(i)
                 print(row_format.format(i, relay_name, session.get_relay_position(relay_name), session.get_relay_count(relay_name)))
+
```
Comment (Fladolcetta) [inline src/niswitch/examples/niswitch_get_device_info.py]: Whats the policy for spaces between try catch blocks?

File: src/niswitch/examples/niswitch_get_device_info.py [hunk 17]
```diff
@@ -4,26 +4,44 @@
...
+            print(row_format.format('Relay count: ', session.number_of_relays))
+            print('Relay Info:')
+
+        if args.channel_info:
+            row_format = '{:6}' + ' ' * 12 + '{:<15}{:<22}{:6}'
+            print(row_format.format('Number', 'Name', 'Is Configuration', 'Is Source'))
+            for i in range(1, session.channel_count + 1):
+                channel_name = session.get_channel_name(i)
+                with session[channel_name] as channel:
+                    print(row_format.format(i, channel_name, str(channel.is_configuration_channel), str(channel.is_source_channel)))
```
Comment (Fladolcetta) [inline src/niswitch/examples/niswitch_get_device_info.py]: Do we need to coerce the booleans to a string? Isn't this already done by print? For relay position, it was converted from enum to string automatically.

File: docs/niswitch/examples.rst [hunk 18]
```diff
@@ -1,6 +1,15 @@
 Examples
 ========
 
+.DS_Store
```
Comment (texasaggie97) [inline docs/niswitch/examples.rst]: You need to remove all non examples from this folder.

Comment (texasaggie97) [reply docs/niswitch/examples.rst]: If you merge #336 into your branch, files that don't end in .py won't be added as an example.

File: src/niswitch/examples/niswitch_get_device_info.py [hunk 19]
```diff
@@ -4,26 +4,44 @@
 import niswitch
 import sys
 
-parser = argparse.ArgumentParser(description='Gets device information from a specified switch module.', formatter_class=argparse.ArgumentDefaultsHelpFormatter)
+parser = argparse.ArgumentParser(description='Prints information for the specified National Instruments Switch module.', formatter_class=argparse.ArgumentDefaultsHelpFormatter)
 parser.add_argument('-n', '--name', default='PXI1Slot2', help='Resource name of a National Instruments Switch.')
-parser.add_argument('-i', '--information', default=False, action='store_true', help='Provides relay name, position, and count for all relays on a switch')
+parser.add_argument('-d', '--device-info', default=False, action='store_true', help='Prints information for the device')
+parser.add_argument('-c', '--channel-info', default=False, action='store_true', help='Prints information for all relays on the device')
+parser.add_argument('-r', '--relay-info', default=False, action='store_true', help='Prints information for all channels on the device')
```
[the remark this answers is not in the document]
Comment (marcoskirsch) [reply src/niswitch/examples/niswitch_get_device_info.py]: ok

File: src/niswitch/examples/niswitch_get_device_info.py (hunk 15 above)
Comment (marcoskirsch) [reply src/niswitch/examples/niswitch_get_device_info.py]: I put it in the right place and added the other two table titles.

File: src/niswitch/examples/niswitch_get_device_info.py (hunk 17 above)
Comment (marcoskirsch) [reply src/niswitch/examples/niswitch_get_device_info.py]: If I don't have it, it prints 1 and 0.

File: src/niswitch/examples/niswitch_get_device_info.py (hunk 16 above)
Comment (marcoskirsch) [reply src/niswitch/examples/niswitch_get_device_info.py]: I don't know of one.
I added some whitespace to make the code "look better" in my subjective opinion. I tried removing them and it still looks fine. The 4 space indentation delta helps.
I'll post it that way.

File: docs/niswitch/examples.rst (hunk 18 above)
Comment (marcoskirsch) [reply docs/niswitch/examples.rst]: will merge

**Comment by marcoskirsch** [comment]
All the feedback has been addressed except for the unit test improvements. I'm working on some big picture arrangements which I will do a separate PR for in a bit.

**Comment by marcoskirsch** [comment]
Once #337 is merged I'll add the updated tests for channel based properties.

**Comment by marcoskirsch** [comment]
All feedback has been addressed. Awaiting your approvals. Thanks!

File: build/templates/session.py.mako [hunk 20]
```diff
@@ -158,3 +138,51 @@ init_call_params = helper.get_params_snippet(init_function, helper.ParamListType
...
+    '''${config['session_class_description']}'''
+
+    def __init__(${init_method_params}):
+        super(Session, self).__init__(repeated_capability='')
+        # TODO(marcoskirsch): private members should start with _
+        self.${config['session_handle_parameter_name']} = 0  # This must be set before calling _init_with_options.
+        self.${config['session_handle_parameter_name']} = self.${init_function['python_name']}(${init_call_params})
+        self._is_frozen = True
+
+    def __enter__(self):
```
[the remark this answers is not in the document]
Comment (marcoskirsch) [reply build/templates/session.py.mako]: That's a fair critique. Please open an issue so we can discuss there.
